File: SFConfigManager/Parsers/SolutionParser.cs
using Microsoft.Build.Construction;
using System.Collections.Immutable;
using System.Linq;

namespace SFConfigManager.Parsers
{
    public class SolutionParser : ILoader
    {
        private ImmutableArray<string> paths;
        private SolutionFile sl;

        public bool LoadFromFile(string path)
        {
            sl = SolutionFile.Parse(path);
            Fill();
            return true;
        }

        private void Fill()
        {
            paths = sl.ProjectsInOrder.Where(p => p.RelativePath.EndsWith("sfproj")).Select(p => p.AbsolutePath).ToImmutableArray();
        }

        public ImmutableArray<string> GetSFProjFilePaths()
        {
            return paths;
        }
    }
}

File: SFConfigManager.Core/Parsers/SolutionParser.fs
module SFConfigManager.Core.Parsers.SolutionParser

open Microsoft.Build.Construction
open FSharpPlus
open SFConfigManager.Data.Parsers.ParserTypes
open System.IO

let private buildResult sfProjs = { SfProjList = sfProjs }

let parseSolution path =
    let absPathGetter (p: ProjectInSolution) = p.AbsolutePath
    let absPath = Path.GetFullPath(path)

    try
        let sfProjs =
            SolutionFile.Parse(absPath).ProjectsInOrder
            |> Seq.map absPathGetter
            |> Seq.filter (String.endsWith "sfproj")
            |> Seq.toList

        buildResult sfProjs |> Ok

    with
    | e -> Error e
